EmailResponder/FeedbackDecryptor/redactor.py

```diff
@@ -0,0 +1,356 @@
+def _generate_feedback_scheme1(client_platform, client_version, msg):
+    return {
+        "Metadata": {
+            "platform": client_platform
+        },
+        "DiagnosticInfo": {
+            "SystemInformation": {
+                "PsiphonInfo": {
+                    "CLIENT_VERSION": client_version
+                },
+            },
+            "DiagnosticHistory": [
+                {
+                    "data": {
+                        "msg": msg,
+                    },
+                    "msg": msg,
+                    "timestamp": "",
+                }
+            ],
+        },
+    }
+
+
+def _redact_sensitive_values_ios_vpn_test():
+
+    # Test where a sensitive value is redacted
+
+    log = 'ExtensionInfo: {"PacketTunnelProvider":{"Event":"Start","StartMethod":"Container","ExpectFieldToBeRedacted":{"ExpectFieldToBeRedacted":"ExpectValueToBeRedacted"}}}'
+
+    obj = _generate_ios_vpn_feedback(171, log)
+    redact_sensitive_values(obj)
+
+    # The order of fields in the JSON string may change due to reserialization # during redaction.
+    expectedOutputOrdering1 = _generate_ios_vpn_feedback(171, 'ExtensionInfo: {"PacketTunnelProvider": {"Event": "Start", "StartMethod": "Container"}}')
+    expectedOutputOrdering2 = _generate_ios_vpn_feedback(171, 'ExtensionInfo: {"PacketTunnelProvider": {"StartMethod": "Container", "Event": "Start"}}')
+
+    assert(obj == expectedOutputOrdering1 or
+           obj == expectedOutputOrdering2)
+    
+    # Test where no redaction attempts are made based on the client version
+
+    obj = _generate_ios_vpn_feedback(1, log)
+    obj_copy = _generate_ios_vpn_feedback(1, log)
+    redact_sensitive_values(obj)
+    assert(obj == obj_copy)
+
+    # Test where no redaction attempts are made based on the client platform
+
+    obj = _generate_android_feedback(171, log)
+    obj_copy = _generate_android_feedback(171, log)
+    redact_sensitive_values(obj)
+    assert(obj == obj_copy)
+
+    print 'redact_sensitive_values_ios_vpn_test okay'
+
+
+def redactors(client_platform, client_version):
+    '''
+    Return redactors to use for the target client platform and version.
+    '''
+    redactors = [_redact_upstream_proxy_errors]
+    if client_platform == "ios-vpn" and client_version >= 160:
+        redactors.append(_ios_vpn_redact_start_tunnel_with_options)
+    return redactors
+
+
+def run_redactors(obj, redactors):
+    '''
+    Traverse the object and perform any necessary redactions.
+    '''
+    for path, val in utils.objwalk(obj):
+        for redactor in redactors:
+            redactor(obj, path, val)
+
+diagnostic_msg_regex = re.compile(r'([a-zA-Z]+): ({.*})')
+
+
+def _redact_upstream_proxy_errors(obj, path, val):
+    '''
+    Redacts any text which follows the target upstream proxy error string.
+    '''
+    if isinstance(val, utils.string_types):
+
+        target = "upstreamproxy error: proxyURI url.Parse: parse "
+
+        index = val.find(target)
```

Comment:
This is probably fine, but you should be aware of (and probably comment about) the risks of searching for text in JSON. It's valid JSON to have `"\u0061\u0062\u0063"` to encode `"abc"`, and you'll get `abc` after decoding, but you won't find it by grepping the JSON. _Probably_ we just "know" that our clients don't encode in ridiculous ways, so you don't need JSON-parse first, but you should comment.

Another problem would be if you had a character in your target that must be encoded to escaped in JSON -- like `\"`. Again, you wouldn't be able to find your literal string in the JSON. You don't, but if someone in the future copied the style of this function they might have a problem. Again, a comment would help.

This warning also applies to the regex below.

(Note that a weird attacker could craft JSON that would bypass your checks. But... why? That's not what we're defending against here.)